### PR TITLE
Scored colour filter

### PR DESCRIPTION
**Pull request by jamieparkinson**
The current implementation of the images colour filter is applied in the ES filter context and so does not take into account the scoring of the colour query. This means that the results can, confusingly, show very good colour palette matches much further down than poorer matches. This PR moves the filter into the query context (inside a `bool.must`) and so affects the scoring.

### CODE REVIEW

File: api/api/src/main/scala/uk/ac/wellcome/platform/api/services/ElasticsearchService.scala [hunk 1]
```diff
@@ -48,7 +50,7 @@ class ElasticsearchService(elasticClient: ElasticClient)(
       .request(
         queryOptions,
         index,
-        scored = queryOptions.searchQuery.isDefined
+        scored = queryOptions.searchQuery.isDefined || queryOptions.scoredFilters.nonEmpty
```
Comment (jamesgorrie) [inline api/api/src/main/scala/uk/ac/wellcome/platform/api/services/ElasticsearchService.scala:53]: It doesn't feel like it makes that much sense to have this on a filter per filter bases, and might lead to confusion.

Why not have it as an explicit value. on the `ElasticsearchQueryOptions`?

Comment (jamesgorrie) [reply api/api/src/main/scala/uk/ac/wellcome/platform/api/services/ElasticsearchService.scala:53]: The reason it feels odd is it is on a per request basis, and it feels hidden within the filters.

File: api/api/src/main/scala/uk/ac/wellcome/platform/api/services/ImagesRequestBuilder.scala [hunk 2]
```diff
@@ -36,9 +37,12 @@ class ImagesRequestBuilder(queryConfig: QueryConfig)
...
           .getOrElse(boolQuery)
-          .filter(queryOptions.filters.collect {
-            case filter: ImageFilter => buildImageFilterQuery(filter)
-          })
+          .must(
+            buildImageFilterQuery(queryOptions.scoredFilters)
+          )
+          .filter(
+            buildImageFilterQuery(queryOptions.unscoredFilters)
+          )
```
Comment (jamesgorrie) [inline api/api/src/main/scala/uk/ac/wellcome/platform/api/services/ImagesRequestBuilder.scala:45]: Again feels strange as these are - could we not have `DocumentFilters` and `DocumentQueries`?

File: api/api/src/main/scala/uk/ac/wellcome/platform/api/services/ElasticsearchService.scala (hunk 1 above)
Comment (jamieparkinson) [reply api/api/src/main/scala/uk/ac/wellcome/platform/api/services/ElasticsearchService.scala:53]: I'm not sure I understand the comment about it not making sense on "filter per filter bases" but I agree that having it all live in `filters` / `DocumentFilter` feels odd (it was just the minimal change given how much the query config gets passed around through various case classes etc). That said, this is still a filter because it's a `must`, not a `should`, and so it's hard to know how to name it in such a way that indicates how it should be used...

Are you suggesting that we have `scoredFilters` and `unscoredFilters` as explicit fields that are populated "manually" on `ElasticsearchQueryOptions` and `ImagesSearchOptions`? Or something else?

**Comment by kenoir** [comment]
@jamieparkinson Please could you fill out the description to contextualise the change a bit?

File: api/api/src/main/scala/uk/ac/wellcome/platform/api/services/ElasticsearchService.scala (hunk 1 above)
Comment (jamesgorrie) [reply api/api/src/main/scala/uk/ac/wellcome/platform/api/services/ElasticsearchService.scala:53]: I am concerned as we have created abstractions on top of elastic DSL, which nearly always feel as though they can be expressed as part of the DSL.

Currently we have starting at the request =>
`QueryParams` => `SearchOptions`
`SearchOptions` => `ElasticsearchQueryOptions`
`ElasticsearchQueryOptions` => `ElasticsearchDSL`

I wonder if we had to start with what we want in Elastic i.e. and work up to the `SearchOptions` we might have something similar, if anything we could cut out `SearchOptions` => `ElasticsearchQueryOptions`.

For this instant we might have:
```
scala case SearchOptions(filters: List[Filter], musts:List[Musts], size: Int, sort: List[Sort])
```

Which can then be converted directly into ES queries.

In short, I would do what you've suggested explicitly saying these are `mustQueries` to avoid having to figure out what that infers further down the line.

**Review by jamesgorrie — APPROVED** [review]
Neato!
